js/stdlib/ts/ALGO.ts

```diff
@@ -1188,38 +1192,64 @@ const getApplicationInfoM = async (idn:BigNumber): Promise<OrExn<AppInfo>> => {
     return { val: res };
   } catch (e:any) {
     debug(dhead, 'node err', e);
-    if ( e?.response?.body?.message === 'application does not exist' ) {
-      return { exn: e };
-    }
   }
+
+  // If algod couldn't find it, lookup application in indexer
   const indexer = await getIndexer();
-  const q = indexer.lookupApplications(id) as unknown as ApiCall<IndexerAppInfoRes>;
-  const res = await doQueryM_(dhead, q);
-  debug(dhead, 'indexer', res);
-  return 'exn' in res ? res : { val: res.val.application };
+  const query = indexer.lookupApplications(id)
+                       .includeAll(true) as unknown as ApiCall<IndexerAppInfoRes>;
+  const queryRes = await doQueryM_(dhead, query);
+
+  if ('val' in queryRes) {
+    // If application was deleted, synthesize AppInfo from transaction data
+    return queryRes.val.application.deleted ? getDeletedApplicationInfoM(id)
+                                            : { val: queryRes.val.application };
+  } else {
+    return queryRes;
+  }
 };
 
-const getDeletedApplicationInfoM = async (idn: BigNumber): Promise<OrExn<AppInfo>> => {
+const getDeletedApplicationInfoM = async (id: number): Promise<OrExn<AppInfo>> => {
   const dhead = 'getDeletedApplicationInfoM'
   const indexer = await getIndexer();
   const query = indexer.searchForTransactions()
                        .txType('appl')
-                       .applicationID(bigNumberToNumber(idn))
-                       .limit(1);
+                       .applicationID(id)
+                       .limit(1) as unknown as ApiCall<IndexerQueryMRes>;
   const queryRes = await doQueryM_(dhead, query);
 
   if ('val' in queryRes) {
+    if (queryRes.val.transactions.length === 0) {
+      return { exn: 'application does not exist' };
+    }
+
     const txn = queryRes.val.transactions[0];
     const appTxn = txn['application-transaction'];
     debug(dhead, {txn});
+
+    if (appTxn === undefined
+        || txn['created-application-index'] === undefined
+        || appTxn['application-id'] !== BigInt(0)
+        || appTxn['approval-program'] === undefined
+        || appTxn['clear-state-program'] === undefined
+        || appTxn['local-state-schema'] === undefined
+        || appTxn['global-state-schema'] === undefined)
+    {
+      return { exn: 'sanity check failed' };
```

Comment:
I don't like this error... I think it should say something like "I thought it was deleted, but when I went to go get the deployment transaction it wasn't right"